subprojects/build-cache-packaging/src/main/java/org/gradle/caching/internal/command/BuildCacheCommandFactory.java

```diff
@@ -121,8 +125,13 @@ public OriginMetadata getOriginMetadata() {
                 List<FileSystemSnapshot> roots = new ArrayList<>();
 
                 if (treeSnapshot == null) {
-                    fileSystemMirror.putMetadata(internedAbsolutePath, DefaultFileMetadata.missing());
-                    fileSystemMirror.putSnapshot(new MissingFileSnapshot(internedAbsolutePath, root.getName()));
+                    MissingFileSnapshot missingFileSnapshot = new MissingFileSnapshot(internedAbsolutePath, root.getName());
+                    if (virtualFileSystem == null) {
+                        fileSystemMirror.putMetadata(internedAbsolutePath, DefaultFileMetadata.missing());
+                        fileSystemMirror.putSnapshot(missingFileSnapshot);
+                    } else {
+                        virtualFileSystem.updateWithKnownSnapshot(internedAbsolutePath, missingFileSnapshot);
+                    }
```

Comment:
It would be great if we could add a `VirtualFileSystem` implementation that delegates back to the `FileSystemMirror` instead of adding `if`s all over the place.

Reply:
That's what we are doing now.